### PR TITLE
chore: e2e tests for reduce. Closes #338

### LINKED ISSUE #338: End to End test for Reduce
- [ ] Non-keyed stream
- [ ] Keyed stream
- [ ] A complex pipeline with multiple non-keyed and key stream together

### PR DESCRIPTION
**Pull request by yhl25**
Signed-off-by: Yashash H L <yashashhl25@gmail.com>

Closes #338

### CODE REVIEW

**Review by vigith — APPROVED** [review]

**Comment by whynowy** [comment]
@yhl25 - could you resolve the conflicts?

**Comment by yhl25** [comment]
I am seeing some data loss with multiple reduce vertices, I am trying to reproduce and debug that

**Comment by yhl25** [comment]
> I am seeing some issues with multiple reduce vertices, I am trying to reproduce and debug that

fixed https://github.com/numaproj/numaflow/pull/402

File: test/reduce-e2e/reduce_test.go [hunk 1]
```diff
@@ -31,8 +31,119 @@ type ReduceSuite struct {
...
+			Expect().
+			Status(204)
+		HTTPExpect(r.T(), "https://localhost:8443").POST("/vertices/in").WithBytes([]byte("3")).WithHeader("X-Numaflow-Event-Time", eventTime).
+			Expect().
+			Status(204)
+	}
+
+	// since the key can be even or odd and the window duration is 10s
+	// the sum should be 20(for even) and 40(for odd)
+	//w.Expect().VertexPodLogContains("sink", "20")
```
Comment (vigith) [inline test/reduce-e2e/reduce_test.go:68]: remove this line

File: test/reduce-e2e/reduce_test.go [hunk 2]
```diff
@@ -31,8 +31,119 @@ type ReduceSuite struct {
...
+			Status(204)
+	}
+
+	// since the key can be even or odd and the first window duration is 10s(which is keyed)
+	// and the second window duration is 60s(non-keyed)
+	// the sum should be 180(60 + 120)
+	w.Expect().VertexPodLogContains("sink", "Payload -  180  Key -  NON_KEYED_STREAM  Start -  120000  End -  180000")
+}
+
+func (r *ReduceSuite) TestSimpleReducePipelineFailOver() {
```
Comment (vigith) [inline test/reduce-e2e/reduce_test.go:145]:
```suggestion
func (r *ReduceSuite) TestSimpleReducePipelineFailOverUsingWAL() {
```

**Review by vigith — APPROVED** [review]

File: test/reduce-e2e/reduce_test.go [hunk 3]
```diff
@@ -31,8 +31,118 @@ type ReduceSuite struct {
...
+
+	// since the key can be even or odd and the first window duration is 10s(which is keyed)
+	// and the second window duration is 60s(non-keyed)
+	// the sum should be 180(60 + 120)
+	w.Expect().VertexPodLogContains("sink", "Payload -  180  Key -  NON_KEYED_STREAM  Start -  120000  End -  180000")
+}
+
+func (r *ReduceSuite) TestSimpleReducePipelineFailOverUsingWAL() {
+	w := r.Given().Pipeline("@testdata/simple-reduce-pipeline-wal.yaml").
 		When().
```
Comment (yhl25) [inline test/reduce-e2e/reduce_test.go:146]: @xdevxy `TestSimpleReducePipelineFailOverUsingWAL` test keeps failing, I have increased the number of input messages, but I am not sure if that will fix the issue. Can you please check?

Comment (yhl25) [reply test/reduce-e2e/reduce_test.go:146]: looks like increasing the number of input messages has fixed it.

Comment (vigith) [reply test/reduce-e2e/reduce_test.go:146]: yes, we need to figure out "the why"

Comment (vigith) [reply test/reduce-e2e/reduce_test.go:146]: opened #408

File: test/fixtures/expect.go [hunk 4]
```diff
@@ -99,8 +99,9 @@ func (t *Expect) VertexPodLogContains(vertexName, regex string, opts ...PodLogCh
 		t.t.Fatalf("Failed to check vertex %q pod logs: %v", vertexName, err)
 	}
 	if !contains {
-		t.t.Fatalf("Expected vertex %q pod log contains %q", vertexName, regex)
+		t.t.Fatalf("Expected vertex %q pod log doesn't contain %q", vertexName, regex)
```
Comment (ashwinidulams) [inline test/fixtures/expect.go:102]:
```suggestion
		t.t.Fatalf("Expected vertex %q pod log to contain %q but didn't.", vertexName, regex)
```

**Review by ashwinidulams — APPROVED** [review]

**Pull request merged**